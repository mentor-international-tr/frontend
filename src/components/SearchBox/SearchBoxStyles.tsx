import styled from "styled-components";
import vars from "lib/styles/variables";
import { InputBase, Paper } from "@material-ui/core";

const MIPaper = styled(Paper)`
  display: flex;
  justify-content: center;
  max-width: 600px;
  border: ${(props: { ismicon: string }) =>
    props.ismicon === "true"
      ? `2px solid ${vars.light_blue}`
      : ""};
`;

const MIInputBase = styled(InputBase)`
  width: 100%;
`;

export { MIPaper, MIInputBase };
