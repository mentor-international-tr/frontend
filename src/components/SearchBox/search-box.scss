@import "../../lib/styles/variables.scss";
.search-container {
    display: flex;
    justify-content: center;
    max-width: 600px;
}

.search-container__search-box {
    width: 100%;
}

.search-container__mic--on {
    border: 2px solid $light-blue;
}
