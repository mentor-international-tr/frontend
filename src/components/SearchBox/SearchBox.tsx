import { MIPaper, MIInputBase } from "./SearchBoxStyles";
import React from "react";
import { IconButton } from "@material-ui/core";

import SpeechRecognition from "react-speech-recognition";
import { useState } from "react";
export interface SearchBoxProps {
  setSearchText: Function;
  searchText: Function;
  searchHandler: Function;
  placeholderWhileMicOn: string;
  placeholder: string;
  micAutoStart: boolean;
  onMicStartHandler?: Function;
  onMicEndHandler?: Function;
  startListening: Function;
  abortListening: Function;
  transcript: string;
  resetTranscript: Function;
  browserSupportsSpeechRecognition: boolean;
}

const SearchBox: React.SFC<SearchBoxProps> = ({
  setSearchText,
  searchText,
  searchHandler,
  onMicStartHandler,
  onMicEndHandler,
  placeholder,
  placeholderWhileMicOn,
  micAutoStart,
  startListening,
  abortListening,
  transcript,
  resetTranscript,
  browserSupportsSpeechRecognition,
}) => {
  // console.log(transcript);
  const [isMicOn, setIsMicOn] = useState(micAutoStart);

  return (
    <MIPaper
      elevation={1}
      ismicon={isMicOn ? "true" : "false"}
    >
      <IconButton
        aria-label="Search"
        data-testid="search-icon-button"
        onClick={() => {
          searchHandler();
        }}
      >
        <i className="fas fa-search" />
      </IconButton>
      <MIInputBase
        placeholder={`${
          isMicOn ? placeholderWhileMicOn : placeholder
        }`}
        className="search-container__search-box"
        data-testid="search-box"
        value={
          isMicOn
            ? `${searchText} ${transcript}`
            : searchText
        }
        onChange={(e: any) => {
          if (isMicOn) {
            setSearchText(`${searchText} ${transcript}`);
          } else {
            setSearchText(e.target.value);
          }
        }}
        onKeyDown={(e: any) => {
          if (e.keyCode === 13) {
            searchHandler();
          }
        }}
      />
      {browserSupportsSpeechRecognition && !isMicOn ? (
        <IconButton
          aria-label="Search By Microphone"
          data-testid="mic-icon-button"
          onClick={() => {
            if (onMicStartHandler) {
              onMicStartHandler();
            }
            startListening();
            setIsMicOn(true);
          }}
        >
          <i className="fas fa-microphone" />
        </IconButton>
      ) : null}
      {browserSupportsSpeechRecognition && isMicOn ? (
        <IconButton
          aria-label="Clear search box and exit microphone mode"
          data-testid="x-icon-button"
          onClick={() => {
            if (onMicEndHandler) {
              onMicEndHandler();
            }
            setSearchText(`${searchText} ${transcript}`);
            abortListening();
            setIsMicOn(false);
            resetTranscript();
          }}
        >
          <i className="fas fa-times" />
        </IconButton>
      ) : null}
    </MIPaper>
  );
};
SearchBox.defaultProps = {
  micAutoStart: false,
  placeholder: "",
  placeholderWhileMicOn: "",
};
export default SpeechRecognition({ autoStart: false })(
  SearchBox,
);
