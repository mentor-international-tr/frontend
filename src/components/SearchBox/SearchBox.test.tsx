import React from "react";
import { render, cleanup, fireEvent } from "react-testing-library";
import SearchBox from "./SearchBox";
import "jest-dom/extend-expect";

afterEach(cleanup);
describe("test <SearchBox />", () => {
  it("creates snapshots", () => {
      //snapshot of component without browser support for speech recognition
    const { asFragment, rerender } = render(
      <SearchBox
        setSearchText={() => {}}
        searchText={""}
        searchHandler={() => {}}
        onMicStartHandler={() => {}}
        onMicEndHandler={() => {}}
      />,
    );
    //snapshot of component with browser supporting speech recognition
    expect(asFragment()).toMatchSnapshot();
    rerender(
        <SearchBox
          setSearchText={() => {}}
          searchText={""}
          searchHandler={() => {}}
          onMicStartHandler={() => {}}
          onMicEndHandler={() => {}}
          browserSupportsSpeechRecognition={true}
        />,
      );
      expect(asFragment()).toMatchSnapshot();
      //TODO - add a snapshot when isMicOn is true and browserSupport is true
  });

  it("tests the value of the input search box", () => {
    const { getByTestId, rerender } = render(
      <SearchBox
        setSearchText={() => {}}
        searchText={""}
        searchHandler={() => {}}
        onMicStartHandler={() => {}}
        onMicEndHandler={() => {}}
      />,
    );
    expect(
    (getByTestId("search-box").querySelector("input") as HTMLInputElement)
        .value,
    ).toBe("");

    rerender(<SearchBox
        setSearchText={() => {}}
        searchText={"abc"}
        searchHandler={() => {}}
        onMicStartHandler={() => {}}
        onMicEndHandler={() => {}}
      />)
      expect(
        (getByTestId("search-box").querySelector("input") as HTMLInputElement)
            .value,
        ).toBe("abc");
  });

  it("when search button has been clicked on, it executes the searchHandler once", () => {
    const searchHandler = jest.fn();
    const { getByTestId } = render(<SearchBox
        setSearchText={() => {}}
        searchText={""}
        searchHandler={searchHandler}
        onMicStartHandler={() => {}}
        onMicEndHandler={() => {}}
      />,)

      fireEvent.click(getByTestId("search-icon-button"));
      expect(searchHandler).toHaveBeenCalledTimes(1);
  });
  it("when microphone button has been clicked on once, onMicStartHandler should be called once", () => {
    const startMicHandler = jest.fn();
    const { getByTestId } = render(<SearchBox
        setSearchText={() => {}}
        searchText={""}
        searchHandler={() => {}}
        onMicStartHandler={startMicHandler}
        onMicEndHandler={() => {}}
        browserSupportsSpeechRecognition={true}
      />,)

      fireEvent.click(getByTestId("mic-icon-button"));
      expect(startMicHandler).toHaveBeenCalledTimes(1);
  });
//   it("when microphone button has been clicked twice, onMicStartHandler and onMicEndHandler should be called once", () => {
//     const startMicHandler = jest.fn();
//     const endMicHandler = jest.fn();
//     const { getByTestId } = render(<SearchBox
//         setSearchText={() => {}}
//         searchText={""}
//         searchHandler={() => {}}
//         onMicStartHandler={startMicHandler}
//         onMicEndHandler={() => endMicHandler
//       />,)

//       fireEvent.click(getByTestId("mic-icon-button"));
//       expect(startMicHandler).toHaveBeenCalledTimes(1);
//   });
});
