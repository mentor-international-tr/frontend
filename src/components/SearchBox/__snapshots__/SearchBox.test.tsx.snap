// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <SearchBox /> creates snapshots 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root-1 MuiPaper-elevation1-4 MuiPaper-rounded-2 search-container "
  >
    <button
      aria-label="Search"
      class="MuiButtonBase-root-34 MuiIconButton-root-28"
      data-testid="search-icon-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label-33"
      >
        <i
          class="fas fa-search"
        />
      </span>
      <span
        class="MuiTouchRipple-root-54"
      />
    </button>
    <div
      class="MuiInputBase-root-37 search-container__search-box"
      data-testid="search-box"
    >
      <input
        class="MuiInputBase-input-47"
        placeholder="Enter Organization Name"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`test <SearchBox /> creates snapshots 2`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root-1 MuiPaper-elevation1-4 MuiPaper-rounded-2 search-container "
  >
    <button
      aria-label="Search"
      class="MuiButtonBase-root-34 MuiIconButton-root-28"
      data-testid="search-icon-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label-33"
      >
        <i
          class="fas fa-search"
        />
      </span>
      <span
        class="MuiTouchRipple-root-54"
      />
    </button>
    <div
      class="MuiInputBase-root-37 search-container__search-box"
      data-testid="search-box"
    >
      <input
        class="MuiInputBase-input-47"
        placeholder="Enter Organization Name"
        type="text"
        value=""
      />
    </div>
    <button
      aria-label="Search By Microphone"
      class="MuiButtonBase-root-34 MuiIconButton-root-28"
      data-testid="mic-icon-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label-33"
      >
        <i
          class="fas fa-microphone"
        />
      </span>
      <span
        class="MuiTouchRipple-root-54"
      />
    </button>
  </div>
</DocumentFragment>
`;
